rules/dispatch.go

```diff
@@ -387,3 +333,193 @@ func (r *DefaultRuleRenderer) dispatchChains(
+
+	rootSetMarkRules := make([]Rule, 0)
+	rootFromMarkRules := make([]Rule, 0)
+
+	// start rendering set mark rules.
+	// Since there can be >100 endpoints, putting them in a single list adds some latency to
+	// endpoints that are later in the chain.  To reduce that impact, we build a shallow tree of
+	// chains based on the prefixes of the chains.
+	commonPrefix, prefixes, prefixToNames := r.divideEndpointNamesToPrefixTree(names)
+
+	// Now, iterate over the prefixes.  If there are multiple names in a prefix, we render a
+	// child chain for that prefix.  Otherwise, we render the rule directly to avoid the cost
+	// of an extra goto.
+	var chains []*Chain
+	for _, prefix := range prefixes {
+		ifaceNames := prefixToNames[prefix]
+		logCxt := log.WithFields(log.Fields{
+			"prefix":          prefix,
+			"namesWithPrefix": ifaceNames,
+		})
+		logCxt.Debug("Considering prefix")
+		if len(ifaceNames) > 1 {
+			// More than one name, render a prefix match in the root chain...
+			nextChar := prefix[len(commonPrefix):]
+			ifaceMatch := prefix + "+"
+			childSetMarkChainName := dispatchSetMarkEndpointChainName + "-" + nextChar
+			logCxt := logCxt.WithFields(log.Fields{
+				"childSetMarkChainName": childSetMarkChainName,
+				"ifaceMatch":            ifaceMatch,
+			})
+			logCxt.Debug("Multiple interfaces with prefix, rendering child chain")
+			rootSetMarkRules = append(rootSetMarkRules, Rule{
+				Match: Match().InInterface(ifaceMatch),
+				// Note: we use a goto here, which means that packets will not
+				// return to this chain.  This prevents packets from traversing the
+				// rest of the root chain once we've found their prefix.
+				Action: GotoAction{
+					Target: childSetMarkChainName,
+				},
+			})
+
+			// ...and child chains.
+			childSetMarkEndpointRules := make([]Rule, 0)
+			for _, name := range ifaceNames {
+				logCxt.WithField("ifaceName", name).Debug("Adding rule to child chains")
+
+				childSetMarkEndpointRules = append(childSetMarkEndpointRules, Rule{
+					Match: Match().InInterface(name),
+					Action: GotoAction{
+						Target: EndpointChainName(setMarkPfx, name),
+					},
+				})
+			}
+
+			childSetMarkEndpointChain := &Chain{
+				Name:  childSetMarkChainName,
+				Rules: childSetMarkEndpointRules,
+			}
+
+			chains = append(chains, childSetMarkEndpointChain)
+
+		} else {
+			// Only one name with this prefix, render rules directly into the root
+			// chains.
+			ifaceName := ifaceNames[0]
+			logCxt.WithField("ifaceName", ifaceName).Debug("Adding rule to root chains")
+
+			rootSetMarkRules = append(rootSetMarkRules, Rule{
+				Match: Match().InInterface(ifaceName),
+				Action: GotoAction{
+					Target: EndpointChainName(setMarkPfx, ifaceName),
+				},
+			})
+		}
+	}
+
+	// At the end of set mark chain, set generic endpoint mark. A generic endpoint mark is used when a forward packet
+	// whose incoming interface is neither a workload nor a host endpoint.
+	rootSetMarkRules = append(rootSetMarkRules, Rule{
```

Comment:
I think we need to make sure that we *don't* set this for `caliXXX|tapXXX` interfaces.  `caliXXX` interfaces that we haven't programmed yet shouldn't be allowed to send traffic.

Reply:
Done. Added drop rules to `cali-set-endpoint-mark` chain.